core/job/service/job_service.go

```diff
@@ -381,23 +381,62 @@ func (j *JobService) Validate(ctx context.Context, jobTenant tenant.Tenant, jobS
 
 func (j *JobService) validateDeleteJobs(ctx context.Context, jobTenant tenant.Tenant, toDelete []*job.Spec, logWriter writer.LogWriter) error {
 	me := errors.NewMultiError("delete job specs check errors")
+	toDeleteMap := map[job.FullName]bool{}
+	for _, specToDelete := range toDelete {
+		fullName := job.FullNameFrom(jobTenant.ProjectName(), specToDelete.Name())
+		toDeleteMap[fullName] = true
+	}
+
 	for _, jobToDelete := range toDelete {
-		downstreamList, err := j.repo.GetDownstreamByJobName(ctx, jobTenant.ProjectName(), jobToDelete.Name())
+		downstreams, err := j.getDownstreams(ctx, jobTenant.ProjectName(), jobToDelete.Name(), map[job.FullName]bool{})
 		if err != nil {
 			logWriter.Write(writer.LogLevelError, fmt.Sprintf("[%s] pre-delete check for job %s failed: %s", jobTenant.NamespaceName().String(), jobToDelete.Name().String(), err.Error()))
 			me.Append(err)
 			continue
 		}
+		notDeleted, safeToDelete := isJobSafeToDelete(toDeleteMap, job.DownstreamList(downstreams).GetDownstreamFullNames())
 
-		if len(downstreamList) > 0 {
-			downstreamFullNames := job.DownstreamList(downstreamList).GetDownstreamFullNames()
-			errorMsg := fmt.Sprintf("deletion of job %s will fail. job is being used by %s", jobToDelete.Name().String(), downstreamFullNames.String())
+		if !safeToDelete {
+			errorMsg := fmt.Sprintf("deletion of job %s will fail. job is being used by %s", jobToDelete.Name().String(), job.FullNames(notDeleted).String())
 			logWriter.Write(writer.LogLevelError, fmt.Sprintf("[%s] %s", jobTenant.NamespaceName().String(), errorMsg))
 			me.Append(errors.NewError(errors.ErrFailedPrecond, job.EntityJob, errorMsg))
 			continue
 		}
 	}
-	return errors.MultiToError(me)
+	return me.ToErr()
+}
+
+func isJobSafeToDelete(toDeleteMap map[job.FullName]bool, downstreamFullNames []job.FullName) ([]job.FullName, bool) {
+	notDeleted := []job.FullName{}
+	for _, downstreamFullName := range downstreamFullNames {
+		if _, ok := toDeleteMap[downstreamFullName]; !ok {
+			notDeleted = append(notDeleted, downstreamFullName)
+		}
+	}
+
+	return notDeleted, len(notDeleted) == 0
+}
+
+func (j *JobService) getDownstreams(ctx context.Context, projectName tenant.ProjectName, jobName job.Name, visited map[job.FullName]bool) ([]*job.Downstream, error) {
```

Comment:
instead of renaming `getDownstreams` function, would it be better to rename the `GetDownstream` with `GetDirectDownstream`?

Reply:
Agree that it will be clearer if any direct downstream function is renamed `GetDirectDownstream`. However, there are some other cases as well where we are fetching only direct upstream (which is the usual case). So for consistency's sake, we need to rename it to `GetDirectUpstream`.

If it is too much to change, I am also fine if we just rename the `getDownstreams` to `getAllDownstreams` or something similar.

Reply:
got it, will change to `getAllDownstreams` then